src/net.cpp

```diff
@@ -1928,6 +1921,25 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 continue;
             }
 
+            // Do not make automatic outbound connections to addnode peers, to
+            // avoid using our limited outbound slots for them and to ensure
+            // addnode connections benefit from their intended protections.
+            {
+                const auto str_addr{addr.ToStringAddr()};
+                const auto str_addr_port{addr.ToStringAddrPort()};
+
+                LOCK(m_added_nodes_mutex);
+                if (m_added_nodes.size() < 16 // bound the query to a reasonable limit
+                    && std::any_of(m_added_nodes.cbegin(), m_added_nodes.cend(), [&](const auto& a) { return a == str_addr || a == str_addr_port; })) {
```

Comment:
Good catch with this deficiency!

If `m_added_nodes` is changed to `std::unordered_set`, then this can be a simple `m_added_nodes.count() > 0` and `CConnman::AddNode()` and `CConnman::RemoveAddedNode()` could be simplified (and made faster since they are now `O(number of added nodes)` and will become `O(1)`, I guess performance is irrelevant in practice because of the small number of elements).

Maybe this is worth a separate PR since it is distinct from other commits (one commit to change it to `std::unordered_set` and one to add this check).